### ISSUE TITLE
Replicating CoNLL 2003 NER training

### ISSUE THREAD

**Issue by pvcastro** [allenai/allennlp#2219]
**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.3
 - AllenNLP version: I installed from master
 - PyTorch version: 0.4.1

**Question**
What is the exact jsonnet config you used for the CoNLL 2003 NER training, as described in the ELMo paper? I just used the same setup as is in training_config/ner.jsonnet and training_config/ner_elmo.jsonnet and the best validation F1 Score are too high. For plain ner.jsonnet was 94.32% and for ner_elmo.jsonnet was 96%. Maybe I'm interpreting the results wrong, but I believe the dataset is correct, it's the same ones I got from [Lample's LSTM-CRF repository](https://github.com/glample/tagger/tree/master/dataset)

By default the validation metric "+f1-measure-overall" Is the same as the conll validation script, right?

**Comment by matt-peters** [comment]
The validation metric +f1-metric-overall is the same as the CoNLL script, for this task.  We don't have a config file for the exact version in the ELMo paper as that was implemented in tensorflow.  Are you saying that the allennlp version has higher dev F1 then the version we reported in the paper?  It's probably a little more carefully tuned.  Also, note that there is significant variation with the random seed for this dataset, we reported the average F1 (not max as sometimes done).

**Comment by pvcastro** [comment]
OK, thanks. I made a confusion here with eng.testa and eng.testb :sweat_smile: 

eng.testb is the test dataset and eng.testa is the dev dataset, right? 

ner.jsonnet explicitly defines all 3, but it doesn't seem to print any evaluation results on test dataset, only on validation. In order for test dataset to be evaluated, we need to define evaluate_on_test = True explicitly, right? Just leaving the ner.jsonnet as is is not enough, even though it defines all 3 datasets.

**Comment by matt-peters** [comment]
Yes, testa is dev and testb is test.

Recommend best practice is to leave `evaluate_on_test=False`, and instead run `allennlp evaluate` with the best model evaluated on the dev set.

**Comment by pvcastro** [comment]
OK, great! Even if it was just a single run, now it got consistent with the reported results in the paper, 92.27% F1 Score.

> Recommend best practice is to leave evaluate_on_test=False, and instead run allennlp evaluate with the best model evaluated on the dev set.

Looks like setting `evaluate_on_test=True` already does exactly this, picking the best model evaluated on the dev set. Any reason why doing it with the evaluate command would be a better approach? 

Thanks!

**Comment by DeNeutoy** [comment]
Because if you continuously train models with different hyperparameters, you will be looking at the test set results every time, which is not good practice. You should perform validation on the _validation_ set and then run the best model once on the test set.

**Comment by pvcastro** [comment]
I see that @DeNeutoy. My point is that evaluate_on_test doesn't cause the validation to switch to the test dataset, it only performs an evaluation on the test dataset once the training finishes, only once. The same thing that is done with the evaluate command.

Thanks!
